include/irr/asset/filters/CSwizzleAndConvertImageFilter.h

```diff
@@ -188,20 +188,50 @@ class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle> : public CIma
 			#endif
 			auto perOutputRegion = [&blockDims,inFormat,outFormat,&state](const CMatchedSizeInOutImageFilterCommon::CommonExecuteData& commonExecuteData, CBasicImageFilterCommon::clip_region_functor_t& clip) -> bool
 			{
-				auto swizzle = [&commonExecuteData,&blockDims,inFormat,outFormat,&state](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos)
+				const auto inTexelOrBlockByteSize = asset::getTexelOrBlockBytesize(inFormat);
+				const auto inChannelAmount = asset::getFormatChannelCount(inFormat);
+
+				// TESTING
+				auto getClampedSourcePixels = [&](auto sourcePixels)
+				{
+					constexpr auto MaxPlanes = 4;
+					auto buffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(blockDims.y * blockDims.x * inTexelOrBlockByteSize);
+					const void* newSourcePixels[MaxPlanes] = { buffer->getPointer(), nullptr, nullptr, nullptr };
+
+					for (auto blockY = 0u; blockY < blockDims.y; blockY++)
+						for (auto blockX = 0u; blockX < blockDims.x; blockX++)
+						{
+							std::array<double, MaxPlanes> decodeBuffer = {};
+							asset::decodePixelsRuntime(inFormat, sourcePixels, decodeBuffer.data(), blockX, blockY);
+
+							for (auto& value : decodeBuffer)
+								core::clamp<double, double>(value, 0, 1);
+
+							asset::encodePixelsRuntime(inFormat, reinterpret_cast<uint8_t*>(buffer->getPointer()) + (blockY * blockDims.x + blockDims.x) * inTexelOrBlockByteSize, decodeBuffer.data());
+						}
+
+					return buffer;
+				};
+
+				auto swizzle = [&commonExecuteData,&blockDims,inFormat,outFormat,&state,&getClampedSourcePixels](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos)
 				{
 					constexpr auto MaxPlanes = 4;
 					const void* srcPix[MaxPlanes] = { commonExecuteData.inData+readBlockArrayOffset,nullptr,nullptr,nullptr };
 
+					auto clampedBuffer = getClampedSourcePixels(srcPix); // todo - compile time, do not waste memory
```

Comment:
use `if constexpr` instead of a lambda, also clamp the pixels on output not input.

Reply:
adjusted in following commit

Reply:
wait, probably got one thing wrong, but I will show it when pushed